File: ipcheck/ipcheck.pl
#!/usr/bin/perl -w

use strict;
use warnings;
use Config::IniFiles;
use Getopt::Long;
Getopt::Long::Configure( "no_auto_abbrev" );

use lib '/root/scripts/utils';
use Mail;

# Get the configuration option
my %opts;

GetOptions (
    \%opts,
    "help|h", # This option will display a short help message.
    "config|c:s",   # This option indicates the configuration file to use
);

# Check if the configuration file exists
unless ($opts{'config'})
{
    print "No configuration file passed! Use the -c/--config option to pass a";
    print " configuration file to the script\n";
    exit 1;
}

unless (-r $opts{'config'})
{
    print "The configuration file passed (".$opts{'config'}.") is not ";
    print "readable. Please specify another one\n";
    exit 1;
}

# Get the configuration file
my $cfg = Config::IniFiles->new ( -file => $opts{'config'} );

# Read the configuration
my $logfile = $cfg->val( "General", "Logfile" );
my $old_ip_file = $cfg->val( "General", "OldIpFile" );
my $mail_address = $cfg->val( "Mail", "To" );

# Open the logfile
unless (open( LOG, ">>$logfile" ))
{
    print "Cannot open logfile $logfile!\n";
    exit 1;
}


# Get the date
my $date = `date +"%Y-%m-%dT%H:%M:%S"`;
chomp( $date );

# Get the current IP
my $current_ip = `nslookup patklaey.internet-box.ch | awk -F': ' 'NR==6 { print \$2 } '`;
chomp($current_ip);

if( ! defined $current_ip || $current_ip eq "" ) {
    my $nslookup = `nslookup patklaey.internet-box.ch`;
    print LOG "$date : IP could not be retrieved. Current nslookup result is: $nslookup";
    close LOG;
    exit 0;
}

# Get the old ip address from the file
if (-e $old_ip_file)
{
    open( OLD, $old_ip_file );
    my $old_ip = <OLD>;

    if ($old_ip ne $current_ip)
    {
        # Send the mail
        my $mailer = Mail->new();
        $mailer->setTo($mail_address);
        my $message = "Dear admin, The public IP of our home changed from $old_ip to $current_ip. Please update the bind configuration and switchplus nameserver in order to have a functional DNS.\n\nCheers your ip-check script";
        $message = "Subject: Our Public IP changed to $current_ip\n".$message;
        open( NEW, ">$old_ip_file" );
        print NEW $current_ip;
        close NEW;
        $mailer->send( $message );
        if ($mailer->error())
        {
            print LOG "$date : Mail could not be sent: ".$mailer->error()."\n";
        } else
        {
            print LOG "$date : IP changed from $old_ip to $current_ip. Mail sent";
            print LOG " to the administrator (".join( ",", $mailer->getTo() ).")\n";
        }
    } else
    {
        print LOG "$date : Ipcheck successful, still the same ip: $current_ip";
        print LOG "\n";
    }

} else
{
    open( OLD, ">$old_ip_file" );
    print OLD $current_ip;
    close OLD;
}

close LOG;
exit 0;

File: ipcheck/ipcheck.go
package main

import (
	"bufio"
	"errors"
	"flag"
	"fmt"
	"log"
	"net"
	"os"
	"strings"
	"time"

	"ServerManagement/utils"

	"gopkg.in/yaml.v3"
)

var now string

type configuration struct {
	LogFilePath     string                 `yaml:"logFilePath"`
	OldIpFilePath   string                 `yaml:"oldIpFilePath"`
	PagerDutyConfig *utils.PagerdutyConfig `yaml:"pagerduty"`
}

func main() {
	// ----- CLI flags -----
	configPath := flag.String("config", "", "Path to configuration file")
	flag.StringVar(configPath, "c", "", "Path to configuration file (shorthand)")
	help := flag.Bool("help", false, "Show help")
	flag.BoolVar(help, "h", false, "Show help (shorthand)")
	flag.Parse()

	if *help {
		flag.Usage()
		os.Exit(0)
	}

	if *configPath == "" {
		log.Fatal("No configuration file passed! Use -c or --config")
	}

	if _, err := os.Stat(*configPath); err != nil {
		log.Fatalf("Configuration file %s is not readable: %v", *configPath, err)
	}

	// ----- Load INI config -----
	config := loadConfig(*configPath)
	if config == nil {
		log.Fatalf("Failed to load config file")
		os.Exit(1)
	}

	logfile := config.LogFilePath
	oldIPFile := config.OldIpFilePath

	if logfile == "" || oldIPFile == "" {
		log.Fatal("Missing required configuration values")
	}

	// ----- Open logfile -----
	logFH, err := os.OpenFile(logfile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		log.Fatalf("Cannot open logfile %s: %v", logfile, err)
	}
	defer logFH.Close()

	logger := log.New(logFH, "", 0)

	now = time.Now().Format("2006-01-02T15:04:05")

	// ----- Resolve current IP -----
	currentIP, err := resolveIP("patklaey.internet-box.ch")
	if err != nil {
		logger.Printf("%s : IP could not be retrieved. Error: %v\n", now, err)
		os.Exit(0)
	}

	// ----- Read old IP -----
	oldIP, err := readOldIP(oldIPFile)
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		log.Fatalf("Failed to read old IP file: %v", err)
	}

	if oldIP == "" {
		// First run: write IP
		if err := writeIP(oldIPFile, currentIP); err != nil {
			log.Fatalf("Failed to write IP file: %v", err)
		}
		return
	}

	if oldIP != currentIP {
		options := utils.IncidentOptions{
			BodyDetails: fmt.Sprintf("The public IP of our home changed from %s to %s. Please update the bind configuration and switchplus nameserver in order to have a functional DNS", oldIP, currentIP),
			Title:       "New IP Address",
			Urgency:     utils.UrgencyLow,
		}
		incident, err := utils.CreateIncident(options, config.PagerDutyConfig, logger)
		if err != nil {
			logger.Printf("%s : Failed to create incident: %s\n", now, err)
		} else {
			logger.Printf("%s : Incident successfully created %v\n", now, incident)
		}

		if err = writeIP(oldIPFile, currentIP); err != nil {
			log.Fatalf("Failed to update IP file: %v", err)
		}
	} else {
		logger.Printf("%s : Ipcheck successful, still the same ip: %s\n", now, currentIP)
	}
}

// resolveIP resolves the first IPv4 address for a hostname
func resolveIP(host string) (string, error) {
	ips, err := net.LookupIP(host)
	if err != nil {
		return "", err
	}

	for _, ip := range ips {
		if ipv4 := ip.To4(); ipv4 != nil {
			return ipv4.String(), nil
		}
	}

	return "", errors.New("no IPv4 address found")
}

func readOldIP(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	if scanner.Scan() {
		return strings.TrimSpace(scanner.Text()), nil
	}

	return "", nil
}

func writeIP(path, ip string) error {
	return os.WriteFile(path, []byte(ip), 0644)
}

func loadConfig(configFile string) *configuration {
	var config configuration

	yfile, err := os.ReadFile(configFile)
	if err != nil {
		fmt.Println("Could not read config file: ", err)
		return nil
	}

	err = yaml.Unmarshal(yfile, &config)
	if err != nil {
		fmt.Println("Error loading configuration: ", err)
		return nil
	}
	return &config
}
